### PR TITLE
Makes getFlatIcon slower. (Also more precise)

### LINKED ISSUE #7092: Photographs don't show janicart nor windows
Photographs do not show whole-tile windows.

They also do not capture the janicart.

They also do not get fire lights.

Nor Plasma.

Fix that shit.

### PR DESCRIPTION
**Pull request by AnturK**
Lummox consigned us to getFlatIcon hell so trying to salvage what i can.

Should fix the most common photo issues. (Except space, still working on that)

Quick test for grabbing icons of everything in range 15 of box bridge doesn't show much difference, will have to do overlay heavy (profile humans) test.
Old:
old
New:
new

Fixes #7092

- [ ] Profile this. Ouch.

:cl:
fix: Photos now properly capture everything again.
/:cl:

### CODE REVIEW

**Comment by tortellinitony** [comment]
pick one

**Comment by Iamgoofball** [comment]
can we maybe look into caching this shit or something

**Comment by Cyberboss** [comment]
Fixes #7092?

File: code/__HELPERS/icons.dm [hunk 1]
```diff
@@ -745,10 +755,31 @@ The _flatIcons list is a cache for generated icon files.
 			noIcon = TRUE // Do not render this object.
 
 	var/curdir
-	if(A.dir != 2)
-		curdir = A.dir
-	else
+	var/base_icon_dir	//We'll use this to get the icon state to display if not null BUT NOT pass it to overlays as the dir we have
+	
+	//These should use the parent's direction (most likely)
+	if(!A.dir || A.dir == 2)
```
Comment (Cyberboss) [inline code/__HELPERS/icons.dm]: Why not SOUTH?

File: code/__HELPERS/icons.dm [hunk 2]
```diff
@@ -745,10 +755,31 @@ The _flatIcons list is a cache for generated icon files.
...
+	//Let's check if the icon actually contains any diagonals, just skip if it's south to save (lot of) time
+	if(curdir != SOUTH)
+		var/icon/test_icon 
+		var/directionals_exist = FALSE
+		var/list/dirs_to_check = GLOB.cardinals - SOUTH
+		outer:
+			for(var/possible_dir in dirs_to_check)
+				test_icon = icon(curicon,curstate,possible_dir,frame=1)
+				for(var/x in 1 to 32)
+					for(var/y in 1 to 32)
```
Comment (Cyberboss) [inline code/__HELPERS/icons.dm]: world.icon_size?

File: code/__HELPERS/icons.dm [hunk 3]
```diff
@@ -825,9 +862,9 @@ The _flatIcons list is a cache for generated icon files.
 
 		if(I == copy) // 'I' is an /image based on the object being flattened.
 			curblend = BLEND_OVERLAY
-			add = icon(I.icon, I.icon_state, I.dir)
+			add = icon(I.icon, I.icon_state, base_icon_dir)
 		else // 'I' is an appearance object.
-			add = getFlatIcon(new/image(I), curdir, curicon, curstate, curblend)
+			add = getFlatIcon(new/image(I), curdir, curicon, curstate, curblend, start = FALSE)
```
Comment (Cyberboss) [inline code/__HELPERS/icons.dm]: no need for the named arg

File: code/modules/paperwork/photography.dm [hunk 4]
```diff
@@ -235,7 +227,9 @@
 			blueprints = 1
 
 	for(var/turf/T in turfs)
-		res.Blend(getFlatIcon(T.loc), blendMode2iconMode(T.blend_mode), 32 * (T.x - center.x) + 33, 32 * (T.y - center.y) + 33)
+		var/area/A = T.loc
+		if(A.icon_state)//There's actually something to blend in.
+			res.Blend(getFlatIcon(A), blendMode2iconMode(A.blend_mode), 32 * (T.x - center.x) + 33, 32 * (T.y - center.y) + 33)
```
Comment (Cyberboss) [inline code/modules/paperwork/photography.dm]: more world.icon_size?

File: code/__HELPERS/icons.dm (hunk 2 above)
Comment (AnturK) [reply code/__HELPERS/icons.dm]: Technically icon.width/height but didn't want to add more procalls than necessary.

**Comment by AnturK** [comment]
This should be good to go. Testmerges didn't show as big impact as i feared.

**Comment by Chocobro** [comment]
three years
what a thrill

**Pull request merged**